azavea/echo-locator
# Feature/kak/signup clients#269

## Overview

Adds client accounts that counselors can create from the profile page. Also switch to using email as login username. 

Restricts access to client accounts to their own profile only. Any previously existing counselor-created profile will be migrated to be accessible by the client (and still accessible by counselors) on the client's first login. Hides profile search page, link to search page, and account creation form from client users.


### Demo

The new profile page section for sending invites, with a validation error:
![echo_email_validation](https://user-images.githubusercontent.com/960264/68924366-a678ce80-074e-11ea-98c1-608b678d99cc.png)

The invite section hides after a successful invite, but will reappear on subsequent profile page visits/reloads until the client logs in for the first time.
![echo_invitation_sent](https://user-images.githubusercontent.com/960264/68924405-c5776080-074e-11ea-8bbb-dd0fc4ddfabd.png)

When logged in as a client, or when a counselor views the profile of a client with a confirmed account, the invite section is gone:
![echo_no_invitation_section](https://user-images.githubusercontent.com/960264/68924460-e344c580-074e-11ea-9866-c4a8427ee8c5.png)


### Notes

The staging stack configuration is entirely new. Updating the existing staging stack configuration wasn't possible, due to it being generated with an outdated version of the Amplify CLI. The user pool will need to be recreated anyway, in order to support email login and the new group and custom user property for the voucher number.

Most of the contents in the staging stack directory the CLI generated and can be ignored, apart from the two `app.js` files that define the two functions (Lambdas), one for creating client accounts, and the other for handling profile migration on first client login.

I've added to the staging stack directory a `README.md` with step-by-step instructions for generating the Amplify stack via the CLI and the subsequent modifications needed to the configuration. Once this has merged and staging has been tested, the production Amplify stack will need to be regenerated with the same steps.


## Testing Instructions

 * `cp deployment/amplify/src/aws-exports.js taui/src/aws-exports.js`
 * Create a user account for the new stack if necessary in the [Coginto console](https://console.aws.amazon.com/cognito) and add it to the `counselors` group
 * `cd taui && yarn start`
 * log in to the site as a counselor
 * should still be sent to search page on login
 * create a new profile after an empty search result
 * bottom of profile edit page should have email input and button to invite user
 * email input should validate email
 * invite should be successful with a valid email and replace invite form with success message
 * until first client account login, counselor should be able to resend invitation on revisiting the profile edit page (temporary password is only valid for 24 hours)
 * using invite email temporary password, log in as a client
 * should load the profile the counselor account created
 * should not allow accessing the `/search` page
 * should not show link to search page in map page header
 * profile edit page should not show account invite section
 * log in again as counselor
 * should be able to search for and go to the client's profile still
 * now on the profile edit page, the account invite section should no longer show to the counselor

Closes #269
